service/collector/rate_limit.go

```diff
@@ -0,0 +1,262 @@
+	}
+	if config.Location == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Location must not be empty", config)
+	}
+
+	u := &RateLimit{
+		g8sClient: config.G8sClient,
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+
+		environmentName:          config.EnvironmentName,
+		location:                 config.Location,
+		HostAzureClientSetConfig: config.HostAzureClientSetConfig,
+	}
+
+	return u, nil
+}
+
+func (u *RateLimit) Collect(ch chan<- prometheus.Metric) error {
+	// We need all CRs to gather all subscriptions below.
+	var crs []providerv1alpha1.AzureConfig
+	{
+		mark := ""
+		page := 0
+		for page == 0 || len(mark) > 0 {
+			opts := metav1.ListOptions{
+				Continue: mark,
+			}
+			list, err := u.g8sClient.ProviderV1alpha1().AzureConfigs(metav1.NamespaceAll).List(opts)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			crs = append(crs, list.Items...)
+
+			mark = list.Continue
+			page++
+		}
+	}
+
+	// We generate clients and group them by subscription.
+	// Subscription is defined by the Secret used to fetch the credentials.
+	clientConfigBySubscription := map[string]client.AzureClientSetConfig{}
+	{
+		for _, cr := range crs {
+			config, err := credential.GetAzureConfig(u.k8sClient, key.CredentialName(cr), key.CredentialNamespace(cr))
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			clientConfigBySubscription[config.SubscriptionID] = *config
+		}
+
+		if _, present := clientConfigBySubscription[u.HostAzureClientSetConfig.SubscriptionID]; !present {
+			config := &client.AzureClientSetConfig{
+				ClientID:       u.HostAzureClientSetConfig.ClientID,
+				ClientSecret:   u.HostAzureClientSetConfig.ClientSecret,
+				SubscriptionID: u.HostAzureClientSetConfig.SubscriptionID,
+				TenantID:       u.HostAzureClientSetConfig.TenantID,
+			}
+			clientConfigBySubscription[config.SubscriptionID] = *config
+		}
+	}
+
+	ctx := context.Background()
+
+	// We track RateLimit metrics for each client labeled by SubscriptionID and
+	// ClientID.
+	// That way we prevent duplicated metrics.
+	for _, clientConfig := range clientConfigBySubscription {
+		azureClients, err := client.NewAzureClientSet(clientConfig)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		resourceGroup, err := u.createEmptyResourceGroup(ctx, azureClients)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		// Remaining read requests can be retrieved sending a read request.
+		var reads float64
+		{
+			groupResponse, err := azureClients.GroupsClient.Get(ctx, *resourceGroup.Name)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			reads, err = strconv.ParseFloat(groupResponse.Response.Header.Get(remainingReadsHeaderName), 64)
+			if err != nil {
+				u.logger.Log("level", "warning", "message", "an error occurred parsing to float the value inside the rate limiting header for read requests", "stack", microerror.Stack(microerror.Mask(err)))
+				reads = 0
```

Comment:
Would it make sense to have error counter here and alert on it as a fallback?

Reply:
Done